### ISSUE TITLE
Building mapnik-3.0.4 with boost-1.59 and gcc-5.2.1 fails.

### ISSUE THREAD

**Issue by barjac** [mapnik/mapnik#3072]
Here are the relevant parts of the build log showing the configuration and the final error:

```
----------------snip---------------
Checking for freetype-config... yes
Checking for xml2-config... yes
Checking for dlfcn.h support ... yes
Checking if compiler (c++) supports -std=c++11 flag... (cached) yes
Checking for C library z... yes
Checking for C++ library icuuc... yes
Checking for ICU version >= 4.2... [94mfound: icu 55.1[0m
(cached) Checking for C++ library harfbuzz... yes
Checking for HarfBuzz version >= 0.9.34... [94mfound: HarfBuzz 0.9.41[0m
(cached) Searching for boost libs and headers... (cached) [94m
Found boost libs: /usr/lib64
Found boost headers: /usr/include[0m
Checking for C++ header file boost/version.hpp... yes
Checking for Boost version >= 1.47... yes
[94mFound boost lib version... 1_59[0m
Checking for C++ library boost_system... yes
Checking for C++ library boost_filesystem... yes
Checking for C++ library boost_regex... yes
Checking for C++ library boost_program_options... yes
Checking for C++ library boost_thread... yes
Checking whether Boost was compiled with C++11 scoped enums ... yes
Checking if boost_regex was built with ICU unicode support... (cached) yes
Checking for C library jpeg... yes
Checking for C library proj... yes
Checking for C library png... yes
Checking for C library webp... yes
Checking for C library tiff... yes
Checking for pkg-config... yes
[94mChecking for requested plugins dependencies...[0m
Checking for pg_config... yes
Checking for C library sqlite3... yes
Checking if SQLite supports RTREE... (cached) yes
Checking for gdal-config --libs... yes
Checking for gdal-config --cflags... yes
Checking for name of gdal library... gdal
Checking for C++ library gdal... yes
Checking for pg_config... yes
Checking if gdal is ogr enabled... yes
Checking for gdal-config --libs... yes
Checking for gdal-config --cflags... yes
Checking for name of ogr library... gdal
Checking for C++ library gdal... yes
Checking for cairo... yes
Checking for cairo lib and include paths...  yes
Checking for cairo freetype font support ... yes
[94m
All Required dependencies found!

----------------------snip-------------------

src/json/libmapnik-json.a(mapnik_json_feature_grammar.o): In function `mapnik::util::variant_helper<icu::UnicodeString>::move(unsigned long, void*, void*)':
/home/baz/rpmbuild/BUILD/mapnik-v3.0.4/include/mapnik/util/variant.hpp:221: undefined reference to `icu::UnicodeString::UnicodeString(icu::UnicodeString const&)'
/home/baz/rpmbuild/BUILD/mapnik-v3.0.4/include/mapnik/util/variant.hpp:221: undefined reference to `icu::UnicodeString::UnicodeString(icu::UnicodeString const&)'
src/json/libmapnik-json.a(mapnik_json_feature_grammar.o): In function `mapnik::util::variant<mapnik::value_null, bool, long, double, icu::UnicodeString>::variant<icu::UnicodeString const&, void>(icu::UnicodeString const&)':
/home/baz/rpmbuild/BUILD/mapnik-v3.0.4/include/mapnik/util/variant.hpp:583: undefined reference to `icu::UnicodeString::UnicodeString(icu::UnicodeString const&)'
src/json/libmapnik-json.a(mapnik_json_feature_grammar.o): In function `mapnik::json::attribute_value_visitor::operator()(std::__cxx11::basic_string<char, std::char_traits<char>, std::allocator<char> > const&) const':
/home/baz/rpmbuild/BUILD/mapnik-v3.0.4/include/mapnik/json/feature_grammar.hpp:57: undefined reference to `icu::UnicodeString::~UnicodeString()'
src/json/libmapnik-json.a(mapnik_json_feature_grammar.o): In function `mapnik::util::variant_helper<icu::UnicodeString>::move(unsigned long, void*, void*)':
/home/baz/rpmbuild/BUILD/mapnik-v3.0.4/include/mapnik/util/variant.hpp:221: undefined reference to `icu::UnicodeString::UnicodeString(icu::UnicodeString const&)'
src/json/libmapnik-json.a(mapnik_json_feature_grammar.o): In function `mapnik::util::variant_helper<icu::UnicodeString>::copy(unsigned long, void const*, void*)':
/home/baz/rpmbuild/BUILD/mapnik-v3.0.4/include/mapnik/util/variant.hpp:236: undefined reference to `icu::UnicodeString::UnicodeString(icu::UnicodeString const&)'
collect2: error: ld returned 1 exit status
scons: *** [plugins/input/geojson.input] Error 1
scons: building terminated because of errors.
error: Bad exit status from /home/baz/rpmbuild/tmp/rpm-tmp.bSIUIX (%build)
```

Any help in resolving this would be greatly appreciated!
Barry

**Comment by talaj** [comment]
I suspect you are using G++ 5 and linking against ICU build with old ABI? See https://gcc.gnu.org/onlinedocs/libstdc++/manual/using_dual_abi.html

**Comment by spuhler** [comment]
Hmmm. I rebuilt icu and after mapnik. So they were built against the same libs. Same error.

**Comment by barjac** [comment]
Here is a link to the build log:
http://pkgsubmit.mageia.org/uploads/failure/cauldron/core/release/20150930183516.spuhler.valstar.23629/log/botcmd.1443638164.ecosse.log

**Comment by springmeyer** [comment]
@barjac can you try applying this patch and seeing if the error goes away?

```diff
diff --git a/src/json/build.py b/src/json/build.py
index 390465d..93022b3 100644
--- a/src/json/build.py
+++ b/src/json/build.py
@@ -25,6 +25,7 @@ from glob import glob
 Import('env')
 lib_env = env.Clone()
 lib_env.Append(CXXFLAGS='-fPIC')
+lib_env['LIBS'].append(env['ICU_LIB_NAME'])
 
 name = "mapnik-json"
 lib = lib_env.StaticLibrary(name, glob('./' + '*.cpp'), LIBS=[])
```

**Comment by spuhler** [comment]
Unfortunately, it didn't make a difference. See build log:
http://pkgsubmit.mageia.org/uploads/failure/cauldron/core/release/20150930223949.spuhler.valstar.14607/log/mapnik-3.0.5-1.mga6/build.0.20150930224414.log

**Comment by springmeyer** [comment]
@spuhler - thanks for trying that. I've got a much better idea at a fix now. Can you try:

```diff
diff --git a/plugins/input/geojson/build.py b/plugins/input/geojson/build.py
index 7435be5..60f72f9 100644
--- a/plugins/input/geojson/build.py
+++ b/plugins/input/geojson/build.py
@@ -48,12 +48,12 @@ else:
 
     # Link Library to Dependencies
     libraries = []
-    libraries.append(env['ICU_LIB_NAME'])
     libraries.append('boost_system%s' % env['BOOST_APPEND'])
     libraries.append('mapnik-json')
+    libraries.append(env['ICU_LIB_NAME'])
 
     if env['PLUGIN_LINKING'] == 'shared':
-        libraries.append(env['MAPNIK_NAME'])
+        libraries.insert(0,env['MAPNIK_NAME'])
 
         TARGET = plugin_env.SharedLibrary('../%s' % PLUGIN_NAME,
                                           SHLIBPREFIX='',
```

This should fix the linking of the `geojson.input` which is what is failing in your logs. If this does work then I presume the build will continue on and fail on other plugins which may need the same patch. Let me know what you find.

**Comment by spuhler** [comment]
I am building with it. It will take a while, even with 12 cores

**Comment by spuhler** [comment]
Success, the package compiled. But there is a serious error in the test you may want to look at. This could explain some of the crashes other folks see.
Here is the build log. Ignore the last lines about missing files.
http://pkgsubmit.mageia.org/uploads/failure/cauldron/core/release/20151001005835.spuhler.valstar.20514/log/mapnik-3.0.5-1.mga6/build.0.20151001010313.log

**Comment by springmeyer** [comment]
Thank you @spuhler.

 - original linking error should be fixed now in master (future v3.0.6) after closing #3105.
 - let's track crashing test issues at #3103
